Razor TagHelperDescriptor Hashing

TanayParikh opened the pull request:
1/2 PRs for https://github.com/dotnet/aspnetcore/issues/23170

- I was running into quite a few hash-collisions so decided to revamp the hashing logic throughout the `TagHelperDescriptor` to minimize collisions. 
- Added (collision) testing.
- Added a `TagHelperDescriptor` memory cache.

The other PR utilizes these changes: https://github.com/dotnet/aspnetcore-tooling/pull/2307

TanayParikh committed:
Razor TagHelperDescriptor Hashing and Caching

1/2 PRs for https://github.com/dotnet/aspnetcore/issues/23170

The other PR will be in aspnetcore-tooling to utilize these changes.

TanayParikh commented on src/Shared/RazorShared/TagHelperDescriptorJsonConverter.cs:
```diff
@@ -5,7 +5,6 @@
 using System.Collections.Generic;
 using Microsoft.AspNetCore.Razor.Language;
 using Newtonsoft.Json;
-using Newtonsoft.Json.Linq;
```
Shared file update from [tooling](https://github.com/dotnet/aspnetcore-tooling).

TanayParikh commented on src/Shared/RazorShared/MemoryCache.cs:
```diff
@@ -0,0 +1,72 @@
+    // but until we update O# that causes an Assembly load problem.
+    internal class MemoryCache<TKey, TValue>
+    {
+        private static readonly int SizeLimit = 1500;
```
Large cache size of 1500 `TagHelperDescriptors` done to ensure it works well for large projects (where these changes are most beneficial).

This should be at most around 3MB:

224 serialized taghelpers are 390KB => 390/224*1500/1024 = 2.55 + CacheEntry overhead ~= 3MB.

NTaylorMullen replied:
Given the  reasoning for the size i s based on TagHelpers and you've  made this memory  cache  configurable for  more then just TagHelpers  i'd recommend taking  this  size limit in through the constructor and having  the TagHelperDescriptorCache set it to 1500

TanayParikh replied:
I was thinking of making this TagHelper specific through the following changes:

- SizeLimit of 1500 to start
- At 1500 perform `Compact` _but_ only compact entries which haven't been used in the past ~15 seconds (we already store the `LastAccess` time. If all 1500 entries have been used recently, then expand the size by 500. 
- The purpose of this is we don't have a significant perf drop off at 1500 entries.

Would you say this additional complexity is worthwhile or too much of an edge case?

NTaylorMullen commented on src/Razor/Microsoft.AspNetCore.Razor.Language/src/TagHelperDescriptorCache.cs:
```diff
@@ -0,0 +1,32 @@
+
+namespace Microsoft.AspNetCore.Razor.Language
+{
+    public static class TagHelperDescriptorCache
```
```suggestion
    internal static class TagHelperDescriptorCache
```

NTaylorMullen replied:
Actually, why have this in the language  repo at  all? Seems like  more of a tooling concern given I don't see t his  used  in this PR. Ditto witht he memory cache impl

TanayParikh replied:
Won't that make it package internal to `Microsoft.AspNetCore.Razor.Language` and inaccessible in aspnetcore-tooling based namespaces?

NTaylorMullen commented on src/Razor/Microsoft.AspNetCore.Razor.Language/src/TagHelperDescriptorCache.cs:
```diff
@@ -0,0 +1,32 @@
+        private static readonly MemoryCache<int, TagHelperDescriptor> CachedTagHelperDescriptors = new MemoryCache<int, TagHelperDescriptor>();
+
+        // Disable cache for testing purposes
+        internal static bool CacheEnabled { private get; set; } = true;
```
Lets make it extra clear and  rename  this `CacheEnabledForTest`

TanayParikh committed:
PR Feedback

TanayParikh committed:
Merge branch 'master' into taparik/taghelper_caching

TanayParikh committed:
Update attempt to fix testing missing resource

https://dev.azure.com/dnceng/public/_build/results?buildId=756677&view=ms.vss-test-web.build-test-results-tab&runId=23487428&resultId=102773&paneView=debug

> Manifest resource: Microsoft.AspNetCore.Razor.Language.Test.TestFiles.taghelpers.json not found\nExpected: True\nActual:   False

ajaybhargavb commented on src/Razor/Microsoft.AspNetCore.Razor.Language/src/TagHelperDescriptorComparer.cs:
```diff
@@ -120,8 +120,50 @@ public virtual int GetHashCode(TagHelperDescriptor descriptor)
+            {
+                for (var i = 0; i < descriptor.BoundAttributes.Count; i++)
+                {
+                    hash.Add(descriptor.BoundAttributes[i]);
```
Would it be better to call `GetHashCode()` on each of the bound attributes? Similiarly for other places. I don't know if it makes a difference. Just wondering.

TanayParikh replied:
The `HashCodeCombiner.Add` calls `GetHashCode()` underneath (with appropriate null checks and handling).

TanayParikh committed:
Fix taghelpers.json embedded resource
